src/rapids_pre_commit_hooks/copyright.py

```diff
@@ -0,0 +1,230 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import datetime
+import functools
+import os
+import re
+
+import git
+
+from .lint import LintMain
+
+COPYRIGHT_RE = re.compile(
+    r"Copyright *(?:\(c\))? *(?P<years>(?P<first_year>\d{4})(-(?P<last_year>\d{4}))?),?"
+    r" *NVIDIA C(?:ORPORATION|orporation)"
+)
+BRANCH_RE = re.compile(r"^branch-(?P<major>[0-9]+)\.(?P<minor>[0-9]+)$")
+
+
+class ConflictingFilesError(RuntimeError):
+    pass
+
+
+def match_copyright(content):
+    return list(COPYRIGHT_RE.finditer(content))
+
+
+def strip_copyright(content, copyright_matches):
+    lines = []
+
+    def append_stripped(start, item):
+        lines.append(content[start : item.start()])
+        return item.end()
+
+    start = functools.reduce(append_stripped, copyright_matches, 0)
+    lines.append(content[start:])
+    return lines
+
+
+def apply_copyright_check(linter, old_content):
+    if linter.content != old_content:
+        current_year = datetime.datetime.now().year
+        new_copyright_matches = match_copyright(linter.content)
+
+        if old_content is not None:
+            old_copyright_matches = match_copyright(old_content)
+
+        if old_content is not None and strip_copyright(
+            old_content, old_copyright_matches
+        ) == strip_copyright(linter.content, new_copyright_matches):
+            for old_match, new_match in zip(
+                old_copyright_matches, new_copyright_matches
+            ):
+                if old_match.group() != new_match.group():
+                    if old_match.group("years") == new_match.group("years"):
+                        warning_pos = new_match.span()
+                    else:
+                        warning_pos = new_match.span("years")
+                    linter.add_warning(
+                        warning_pos,
+                        "copyright is not out of date and should not be updated",
+                    ).add_replacement(new_match.span(), old_match.group())
+        else:
+            if new_copyright_matches:
+                for match in new_copyright_matches:
+                    if (
+                        int(match.group("last_year") or match.group("first_year"))
+                        < current_year
+                    ):
+                        linter.add_warning(
+                            match.span("years"), "copyright is out of date"
+                        ).add_replacement(
+                            match.span(),
+                            f"Copyright (c) {match.group('first_year')}-{current_year}"
+                            ", NVIDIA CORPORATION",
+                        )
+            else:
+                linter.add_warning((0, 0), "no copyright notice found")
+
+
+def get_target_branch(repo):
```

Comment:
Let's write a detailed docstring for this function in particular, since it's the most fragile part in my experience. Why this set of candidates for finding branches? What is the intention of each?

Reply:
Added thorough explanation for search algorithm.